components/layout_2020/replaced.rs

```diff
@@ -2,20 +2,77 @@
  * License, v. 2.0. If a copy of the MPL was not distributed with this
  * file, You can obtain one at https://mozilla.org/MPL/2.0/. */
 
+use crate::context::LayoutContext;
 use crate::dom_traversal::NodeExt;
+use crate::formatting_contexts::IndependentLayout;
+use crate::fragments::{Fragment, ImageFragment};
+use crate::geom::{flow_relative, physical};
+use crate::positioned::AbsolutelyPositionedFragment;
+use crate::ContainingBlock;
+use net_traits::image::base::Image;
+use servo_arc::Arc as ServoArc;
+use std::sync::Arc;
 use style::context::SharedStyleContext;
+use style::properties::ComputedValues;
+use style::values::computed::Length;
+use style::Zero;
 
 #[derive(Debug)]
-pub(super) enum ReplacedContent {
-    // Not implemented yet
+pub(crate) struct ReplacedContent {
+    pub kind: ReplacedContentKind,
+    pub intrinsic_size: physical::Vec2<Length>,
+}
+
+#[derive(Debug)]
+pub(crate) enum ReplacedContentKind {
+    Image(Option<Arc<Image>>),
 }
 
 impl ReplacedContent {
-    pub fn for_element<'dom, Node>(element: Node, _context: &SharedStyleContext) -> Option<Self>
-    where
-        Node: NodeExt<'dom>,
-    {
-        // FIXME: implement <img> etc.
+    pub fn for_element<'dom>(element: impl NodeExt<'dom>) -> Option<Self> {
+        if let Some((image, intrinsic_size)) = element.as_image() {
+            return Some(Self {
+                kind: ReplacedContentKind::Image(image),
+                intrinsic_size,
+            });
+        }
         None
     }
+
+    pub fn layout<'a>(
+        &'a self,
+        layout_context: &LayoutContext,
+        style: &ServoArc<ComputedValues>,
+        containing_block: &ContainingBlock,
+    ) -> IndependentLayout {
+        let (fragments, content_block_size) = match self.kind {
+            ReplacedContentKind::Image(ref image) => {
+                // FIXME(nox): We should not assume block size is known.
```

Comment:
Or, on the contrary, isn’t it always known for replaced boxes and should we take `DefiniteContainingBlock` here instead?

Reply:
You are right, at that point we should know both sizes, I'll make it use `DefiniteContainingBlock`.

Reply:
Alright, r=me with that.

Reply:
It turns out we can't do that: this is also called from `IndependentFormattingContext::layout`, where we don't know for sure the block size.

Reply:
Tbh that makes me wonder if we shouldn't separate replaced content from independent formatting contexts.

Reply:
Yeah I considered that when reading your previous message. Or if not separate data types, at least remove `IndependentFormattingContext::layout` and make parent boxes call `IndependentFormattingContext::as_replaced` themselves.

Reply:
That doesn’t need to happen in this PR though. I’ll r+